### PR TITLE
Support FunctionComponent type with no used props

### PR DESCRIPTION
**Pull request by ignatiusreza**
This PR add a failing test case for the issue mentioned in #25

### CODE REVIEW

**Comment by ChristianMurphy** [comment]
hmmm, the core of this is we do know what some props will be ahead of time
https://github.com/rehypejs/rehype-react/blob/39d517894a69157224f96a830055884340d3fbca/types/index.d.ts#L10-L15

if:

```typescript
type ComponentLike<T> = (props: any) => T | null
```

were used `FC` without passing prop types would work, but that comes at the cost of losing type safety.
Something I'd against, since there is a way to allow for using `FC` and keeping the prop types, noted in https://github.com/rehypejs/rehype-react/issues/25#issuecomment-761102136

**Comment by ignatiusreza** [comment]
PR updated, I managed to make it work with `FC`..

the main offender is, `ComponentProps` include `node?: Node` in its type even when `passNode` is not truthy..

by using discriminating union, we can selectively pass in `node` only when `passNode` is `true`..

not sure how to address the issue with class component though.. :thinking:

File: types/index.d.ts [hunk 1]
```diff
@@ -7,14 +7,17 @@ import {Node} from 'unist'
...
+  type ComponentPropsWithoutNode = Record<string, unknown>
+
+  interface ComponentPropsWithNode extends ComponentPropsWithoutNode {
+    node: Node
   }
 
-  type ComponentLike<T> = (props: ComponentProps) => T | null
+  type ComponentProps = ComponentPropsWithoutNode | ComponentPropsWithNode
+
+  type ComponentLike<T, P = ComponentPropsWithNode> = (props: P) => T | null
```
Comment (ChristianMurphy) [inline types/index.d.ts:18]: an unconstrained generic is roughly equivalent to `any`, it doesn't check anything

For example

```typescript
const a: rehypeToReact.ComponentLike<
  ReturnType<typeof React.createElement>,
  // this should require runtime narrowing, but does not
  { notARealProp: Symbol }
> = ({ notARealProp }: { notARealProp: Symbol }) => <>example</>;
```

is valid

this should probably be more along the lines of

```suggestion
type ComponentLike<T, P extends ComponentPropsWithNode = ComponentPropsWithNode> = (props: P) => T | null
```

**Review by ChristianMurphy — APPROVED** [review]
Thanks @ignatiusreza!

File: types/index.d.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Minimum TypeScript Version: 3.4
+// Minimum TypeScript Version: 3.8
```
Comment (ChristianMurphy) [inline types/index.d.ts:1]: TS 3.4 is still supported.
This is bumped to 3.8 to allow adding `// $ExpectError` assertions/test cases on conflicting generic interfaces

File: types/rehype-react-test.tsx [hunk 3]
```diff
@@ -43,6 +48,20 @@ unified().use(rehypeToReact, {
...
+  }
+})
+
+unified().use(rehypeToReact, {
+  createElement: React.createElement,
+  components: {
+    div: ConflictingTypedFunctionalComponent // $ExpectError
+  }
+})
+
```
Comment (ChristianMurphy) [inline types/rehype-react-test.tsx:64]: ensures unsafe implicit type narrowing is not allowed

File: types/rehype-react-test.tsx [hunk 4]
```diff
@@ -52,7 +71,16 @@ unified().use(rehypeToReact, {
...
+    a: (props: rehypeToReact.ComponentPropsWithNode) => <a>{props.node}</a>
+  }
+})
+
+// $ExpectError
+unified().use(rehypeToReact, {
+  createElement: React.createElement,
+  passNode: false,
+  components: {
+    a: (props: rehypeToReact.ComponentPropsWithNode) => <a>{props.node}</a>
```
Comment (ChristianMurphy) [inline types/rehype-react-test.tsx:83]: ensures that the `node` property is only available when `passNode` is active.

**Review by remcohaszing — APPROVED** [review]
I don’t know the exact details, since I don’t use this specific package nor the `React.FC` type, but I don’t see anything that looks wrong and existing type assertions still seem to work.

LGTM :+1:

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
thanks, released!

**Pull request merged**